icu4c/source/i18n/unitsdata.cpp

```diff
@@ -0,0 +1,194 @@
+
+#include <utility>
+
+#include "cstring.h"
+#include "resource.h"
+#include "unitsdata.h"
+#include "uresimp.h"
+
+U_NAMESPACE_BEGIN
+
+namespace {
+
+/**
+ * A ResourceSink that collects conversion rate information.
+ *
+ * This class is for use by ures_getAllItemsWithFallback. Example code for
+ * collecting conversion info for "mile" and "foot" into conversionInfoOutput:
+ *
+ *     UErrorCode status = U_ZERO_ERROR;
+ *     ures_getByKey(unitsBundle, "convertUnits", &fillIn, &status);
+ *     MaybeStackVector<ConversionRateInfo> conversionInfoOutput;
+ *     ConversionRateDataSink convertSink(conversionInfoOutput);
+ *     ures_getAllItemsWithFallback(fillIn, "mile", convertSink, status);
+ *     ures_getAllItemsWithFallback(fillIn, "foot", convertSink, status);
+ */
+class ConversionRateDataSink : public ResourceSink {
+  public:
+    /**
+     * Constructor.
+     * @param out The vector to which ConversionRateInfo instances are to be
+     * added.
+     */
+    explicit ConversionRateDataSink(MaybeStackVector<ConversionRateInfo> &out) : outVector(out) {}
+
+    /**
+     * Adds the conversion rate information found in value to the output vector.
+     *
+     * Each call to put() collects a ConversionRateInfo instance for the
+     * specified source unit identifier into the vector passed to the
+     * constructor, but only if an identical instance isn't already present.
+     *
+     * @param source The source unit identifier.
+     * @param value A resource containing conversion rate info (the base unit
+     * and factor, and possibly an offset).
+     * @param noFallback Ignored.
+     * @param status The standard ICU error code output parameter.
+     */
+    void put(const char *source, ResourceValue &value, UBool /*noFallback*/, UErrorCode &status) {
+        if (U_FAILURE(status)) return;
+        ResourceTable conversionRateTable = value.getTable(status);
+        if (U_FAILURE(status)) return;
+
+        // Collect base unit, factor and offset from the resource.
+        int32_t lenSource = uprv_strlen(source);
+        const UChar *baseUnit = NULL, *factor = NULL, *offset = NULL;
+        int32_t lenBaseUnit, lenFactor, lenOffset;
+        const char *key;
+        for (int32_t i = 0; conversionRateTable.getKeyAndValue(i, key, value); ++i) {
+            if (uprv_strcmp(key, "target") == 0) {
+                baseUnit = value.getString(lenBaseUnit, status);
+            } else if (uprv_strcmp(key, "factor") == 0) {
+                factor = value.getString(lenFactor, status);
+            } else if (uprv_strcmp(key, "offset") == 0) {
+                offset = value.getString(lenOffset, status);
+            }
+        }
+        if (U_FAILURE(status)) return;
+        if (baseUnit == NULL || factor == NULL) {
+            // We could not find a usable conversion rate.
+            status = U_MISSING_RESOURCE_ERROR;
+            return;
+        }
+
+        // Check if we already have the conversion rate in question.
```

Comment:
Why do you need to do this exactly?  To protect yourself from malformed CLDR data?  I don't know why there would be any duplicates in well-formed data.

I see, are you getting the convertUnit for only one unit?  I'm a little confused about the object lifecycles you're trying to use.  To be honest, convertUnit data is pretty small, so maybe you can just load it all at once into static memory and then never have to touch that bundle file again.

Reply:
This is related to Hugo pull request: #31

Reply:
Updated https://github.com/sffc/icu/pull/31 to always grab all conversion units.